### PR TITLE
Added verify-set command to psc-package (issue 2459)

### PR DESCRIPTION
**Pull request by MQuaresma**
Just added, I think, the verify-set command to psc-package as requested. 
PS: If i totatlly missed the point please warn me and I will try to correct it. Thanks!

### CODE REVIEW

File: psc-package/Main.hs [hunk 1]
```diff
@@ -224,6 +224,27 @@ listSourcePaths = do
   paths <- getSourcePaths pkg db depends
   traverse_ (echo . pathToTextUnsafe) paths
 
+verifyPackageSet :: IO ()
+verifyPackageSet  = do
+  -- Clone all repos into the packages/ directory
+  ps <- readPackageFile 
+  paths <- fromList <$>
+    (for (getGitRepoList ps) (\(name, (repo, version)) -> do
```
Comment (paf31) [inline psc-package/Main.hs]: I think you need to add `getGitRepoList`.

File: psc-package/Main.hs [hunk 2]
```diff
@@ -224,6 +224,27 @@ listSourcePaths = do
...
+      let pkgDir = ".psc-package" </> fromText name </> fromText version
```
Comment (paf31) [inline psc-package/Main.hs]: Let's use the usual path here: `.psc-package/<set name>/<package name>/<version>`

File: psc-package/Main.hs [hunk 3]
```diff
@@ -224,6 +224,27 @@ listSourcePaths = do
...
+      exists <- testdir pkgDir
+      unless exists $ cloneShallow repo version pkgDir
+      return (name, pkgDir)))
+
+  -- Print out the psc version
+  procs "psc" [ "--version" ] empty
```
Comment (paf31) [inline psc-package/Main.hs]: We can remove this here actually.

File: psc-package/Main.hs [hunk 4]
```diff
@@ -224,6 +224,27 @@ listSourcePaths = do
...
+
+  for_ (toList ps) $ \(name, PackageSpec{..}) -> do
+    let dirFor = fromMaybe (error "verifyPackageSet: no directory") . (`M.lookup` paths)
+    echo ("Building package " <> name)
+    let srcGlobs = map ((<> "/src/**/*.purs") . toTextUnsafe . dirFor) (name : dependencies)
```
Comment (paf31) [inline psc-package/Main.hs]: Please use `</>` here:

```haskell
"src" </> "**" </> "*.purs"
```

File: psc-package/Main.hs (hunk 1 above)
Comment (MQuaresma) [inline psc-package/Main.hs]: Indeed i do! Sorry for the distraction

File: psc-package/Main.hs [hunk 5]
```diff
@@ -275,7 +296,10 @@ main = do
         , Opts.command "available"
             (Opts.info (pure listPackages)
             (Opts.progDesc "List all packages available in the package set"))
-        ]
+		, Opts.command "verify-all"
```
Comment (paf31) [inline psc-package/Main.hs]: Please fix the indentation here, and add a command description. Also can we change this to `verify-set` please?

**Comment by paf31** [comment]
Thanks for this. Please can you update `contributors.md` too.

**Comment by MQuaresma** [comment]
Ill be sure to review this first thing in the morning tomorrow, it's kind of late here already, Thanks for the help!

**Comment by MQuaresma** [comment]
I think i got it this time, yet again if anything is missing let me know and Ill do my best to correct it!

File: psc-package/Main.hs [hunk 6]
```diff
@@ -224,6 +228,24 @@ listSourcePaths = do
   paths <- getSourcePaths pkg db depends
   traverse_ (echo . pathToTextUnsafe) paths
 
+verifyPackageSet :: IO ()
+verifyPackageSet  = do
+  -- Clone all repos into the packages/ directory
+  ps <- readPackageFile 
+  paths <- fromList <$>
+    (for (getGitRepoList ps) (\(name, (repo, version)) -> do
+      let pkgDir = ".psc-package"  </> fromText repo </> fromText name </> fromText version
```
Comment (paf31) [inline psc-package/Main.hs]: `repo` here should be the package set name. Have you tried running this?

File: psc-package/Main.hs [hunk 7]
```diff
@@ -224,6 +228,24 @@ listSourcePaths = do
...
+      exists <- testdir pkgDir
+      unless exists $ cloneShallow repo version pkgDir
+      return (name, pkgDir)))
+
+  for_ (toList ps) $ \(name, PackageSpec{..}) -> do
+    let dirFor = fromMaybe (error "verifyPackageSet: no directory") . (`M.lookup` paths)
+    echo ("Building package " <> name)
+    let srcGlobs = map ((</>"src"</>"**"</>"*.purs") . toTextUnsafe . dirFor) (name : dependencies)
```
Comment (paf31) [inline psc-package/Main.hs]: Nitpick : please add spaces around the `</>` operators here.

File: psc-package/Main.hs [hunk 8]
```diff
@@ -128,6 +129,9 @@ readPackageSet PackageConfig{ set } = do
       exit (ExitFailure 1)
     Just db -> return db
 
+getGitRepoList :: PackageConfig -> [(PackageName, (Repo, Version))]
+getGitRepoList = toListOf ((ifolded <. to (repo &&& version)) . withIndex)
```
Comment (paf31) [inline psc-package/Main.hs]: This was some code golf while I was working on this originally. I think here, I'd prefer the simpler, lens-free version.

File: psc-package/Main.hs [hunk 9]
```diff
@@ -26,6 +26,7 @@ import qualified Options.Applicative as Opts
 import qualified Paths_purescript as Paths
 import qualified System.IO as IO
 import           Turtle hiding (fold)
+import 			 Control.Lens((<.), toListOf, to, ifolded, withIndex)
```
Comment (LiamGoodacre) [inline psc-package/Main.hs]: These tabs should be spaces. Usually there is a space between the module name and the open parenthesis too. :)

**Comment by paf31** [comment]
It looks like `lens` is missing, but as I noted above, I think it'd be better to try to remove the dependency.

**Comment by MQuaresma** [comment]
@paf31 I dont know if this is common practice but i just wanted to note that Ive haven't been able to work on this issue because of work, hope it's nothing too urgent, ill get back to it as soon as possible.

**Comment by paf31** [comment]
No problem. If you don't have time to finish it, I can finish it off, but there's no huge rush.

**Comment by paf31** [comment]
@MQuaresma I'll close this since it's inactive, but please feel free to reopen it if you are able to finish it off. Thanks!

**Pull request closed without merging**